### PR TITLE
[ENH] Pandas String Methods in a Single Class

### LINKED ISSUE #360: [ENH] Wrap pandas string methods
# Brief Description

One thing that might be nice is to wrap all of the string methods such that they are now method-chainable.

If this could be done programmatically, that would be awesome. Otherwise, manually wrapping ones of interest, and building up a submodule of string method operations, would be a superb alternative.

# Example API

The example API can be found in [this notebook](https://github.com/ericmjl/pyjanitor/blob/dev/examples/notebooks/tidytuesday_20190423.ipynb).

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Single function to access pandas' string methods
- Small subset to see if the current method is acceptable
-

**This PR resolves #360 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [ ] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [ ] Run all tests (`$ pytest .`) locally on your machine.
    - [ ] Check to ensure that test coverage covers the lines of code that you have added.
    - [ ] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [ ] Build the docs locally.
- [ ] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
@samukweku thanks for the PR! I have to admit my brain is struggling a bit to decide whether this goes in `functions.py` or another submodule. Can you share your thoughts here?

The other piece I'm noticing is that some of the names are inconsistent with the string methods. Would it make sense to simply wrap them verbatim?

**Comment by samukweku** [comment]
@ericmjl Thanks for the feedback!
Another submodule is not a bad idea. However, in functions.py we have ``find_replace``, which has a string method attached to it, and ``filter_string`` which uses ``str_contains`` underneath. As such, I am not sure if that would create confusion. 

For the names, the only one is ``concat``, which i felt conveyed what the function was to do, compared to pandas' string ``cat``. But, seeing that  users would be more familiar with the pandas' naming, I'll stick with pandas' names.

The other thought I had, which I want feedback on, applies to chaining. An example could be:

     df.string.split(column='A', pat=',').string.get(2)

Does it look pleasing to the eye? the multiple `string` calls? Dont know if it is something that would put users off.

**Comment by hectormz** [comment]
Is this meant to handle a subset of `pandas.str` methods, or all of them? In the past when I looked at this issue I was intrigued by using a programmatic way to wrap them all

**Comment by samukweku** [comment]
@hectormz Yes, handle all of them. So, just one class to access all string methods

**Comment by hectormz** [comment]
Okay got it. My concern is that there are about 53 `pandas.Series.str` methods.
I'm torn by the manual aspect of it, and would like to see something like the programmatic/generic approach @szuckerman  mentioned.  But it would also need to be obvious and documented enough for users to understand what they're using.

**Comment by ericmjl** [comment]
> @hectormz Yes, handle all of them. So, just one class to access all string methods

I agree with you, @samukweku, this should be the end goal, being able to chain up all of the 

> Okay got it. My concern is that there are about 53 pandas.Series.str methods.
> I'm torn by the manual aspect of it, and would like to see something like the programmatic/generic approach @szuckerman mentioned. But it would also need to be obvious and documented enough for users to understand what they're using.

To make it easier to refer to, [here's all of the string methods][str]. I agree with @hectormz and @szuckerman - an automated thing might be better.

@samukweku, I think we can approach this task in a slightly smarter way than manually wrapping each function manually. This should probably be done over a few PRs, but the general idea probably would involve the following functions:

- A function to identify all pandas string methods, and return their names.
- A wrapper class that checks a function call for its availability as a string method, and executes the function.

From my prior experience, you might end up using the `object.__dict__` representation to make this dynamic passthrough a reality. I'm not sure how deep you'll have to go with calling `object.__dict__["something"].__dict__["somethingelse"]` to consistently passthrough all of the functions, though, so it might take some experimenting. That said, I'm quite sure that this is a supremely fun programming problem!

Now, I just wanted to also mention that even though there may be a more automated way of approaching this task, your work thus far in this PR is still valuable, @samukweku, because it prompted the conversation and realization of how much work it would be to wrap all of the methods manually. So thank you for taking the initiative here!

[str]: https://pandas.pydata.org/docs/user_guide/text.html#method-summary

**Comment by samukweku** [comment]
Thanks for the feedback @ericmjl @hectormz !

**Comment by VPerrollaz** [comment]
- I personally believe that a `string` module might make the methods more discoverable and keep the namespaces tidy.
- From a technical viewpoint, there might also be some useful stuff to pick in [this](https://gist.github.com/VPerrollaz/30d3bce1f7fc678ffd00a19d8231e3fa) for the `__doc__`, `__name__` and `__signature__` dynamic manipulation.
- However this would only help users in ipython/jupyter sessions, I'm not sure how to generate static documentation with sphynx (or anything else actually) from this kind of monkey patching.

**Comment by samukweku** [comment]
Apologies for the delay. Made some updates based on the suggestions. Feedback and suggestions for further improvement. I will add more tests once there is agreement on this function (function name, process, ...)

**Comment by ericmjl** [comment]
Wonderful, wonderful effort, @samukweku! Thank you for handling this. I will be reviewing it this evening, after I finish my work today :).

File: janitor/functions.py [hunk 1]
```diff
@@ -4001,3 +4002,89 @@ def expand_grid(
...
+
+    Functional usage syntax:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor as jn
+
+        df = pd.DataFrame(...)
+        df = jn.process_text(df = df, string_function = "...", args, kwargs)
```
Comment (ericmjl) [inline janitor/functions.py:4058]:
```suggestion
        df = jn.process_text(df = df, string_function = "string_func_name_here", args, kwargs)
```

File: janitor/functions.py [hunk 2]
```diff
@@ -4001,3 +4002,89 @@ def expand_grid(
...
+    Method-chaining usage syntax:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor as jn
+
+        df = (pd.DataFrame(...)
+             .process_text(string_function = "bla", args, kwargs)
+             )
```
Comment (ericmjl) [inline janitor/functions.py:4069]:
```suggestion
        df = (
            pd.DataFrame(...)
            .process_text(string_function = "string_func_name_here", args, kwargs)
        )
```

File: janitor/functions.py [hunk 3]
```diff
@@ -4001,3 +4002,89 @@ def expand_grid(
...
+    """
+
+    data = [
+        func.__name__
+        for _, func in inspect.getmembers(pd.Series.str, inspect.isfunction)
+        if not func.__name__.startswith("_")
+    ]
+
+    if string_function not in data:
+        raise KeyError(f"{string_function} is not a Pandas string method.")
```
Comment (ericmjl) [inline janitor/functions.py:4086]: WOW! :tada:

File: tests/functions/test_process_text.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+def test_str_split():
+    "Test Pandas string split method"
+
+    df = pd.DataFrame(
+        {"text": ["a_b_c", "c_d_e", np.nan, "f_g_h"], "numbers": range(1, 5)}
+    )
+
+    result = process_text(
+        df=df, column="text", string_function="split", pat="_"
+    )
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:18]: @samukweku I think we use the chaining form in the other tests. Unless we haven't been consistent about it? 

```suggestion
    result = df.process_text(
        column="text", string_function="split", pat="_"
    )
```

File: tests/functions/test_process_text.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    "Test Pandas string cat method"
+
+    df = pd.DataFrame({"text": ["a", "b", "c", "d"], "numbers": range(1, 5)})
+
+    result = process_text(
+        df=df,
+        column="text",
+        string_function="cat",
+        others=["A", "B", "C", "D"],
+    )
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:45]:
```suggestion
    result = df.process_text(
        column="text",
        string_function="cat",
        others=["A", "B", "C", "D"],
    )
```

File: tests/functions/test_process_text.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def test_str_get():
+    """Test Pandas string get method"""
+
+    df = pd.DataFrame(
+        {"text": ["aA", "bB", "cC", "dD"], "numbers": range(1, 5)}
+    )
+
+    result = process_text(df=df, column="text", string_function="get", i=-1)
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:61]:
```suggestion
    result = df.process_text(column="text", string_function="get", i=-1)
```

File: tests/functions/test_process_text.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                "john cleese",
+                "terry gilliam",
+                "eric idle",
+                "terry jones",
+                "michael palin",
+            ],
+        }
+    )
+
+    result = process_text(df, column="names", string_function="lower")
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:101]:
```suggestion
    result = df.process_text(column="names", string_function="lower")
```

File: tests/functions/test_process_text.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    assert_frame_equal(result, expected)
+
+
+def test_str_wrong():
+    """Test that string_function is not a Pandas string method"""
+    df = pd.DataFrame(
+        {"text": ["ragnar", "sammywemmy", "ginger"], "code": [1, 2, 3]}
+    )
+    with pytest.raises(KeyError):
+        process_text(df, column="text", string_function="ragnar")
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:112]:
```suggestion
        df.process_text(column="text", string_function="ragnar")
```

File: tests/functions/test_process_text.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    )
+
+    result = process_text(df, column="names", string_function="lower")
+
+    assert_frame_equal(result, expected)
+
+
+def test_str_wrong():
+    """Test that string_function is not a Pandas string method"""
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:107]:
```suggestion
    """Test that string_function that is not a Pandas string method actually errors out"""
```

File: tests/functions/test_process_text.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    expected = pd.DataFrame(
+        {"text": ["A", "B", "C", "D"], "numbers": [1, 2, 3, 4]}
+    )
+
+    assert_frame_equal(result, expected)
+
+
+def test_str_lower():
+    """Test string converts to lowercase"""
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:71]:
```suggestion
    """Test string conversion to lowercase using ``.str.lower()``"""
```

File: tests/functions/test_process_text.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+    expected = pd.DataFrame(
+        {"text": ["aA", "bB", "cC", "dD"], "numbers": [1, 2, 3, 4]}
+    )
+
+    assert_frame_equal(result, expected)
+
+
+def test_str_get():
+    """Test Pandas string get method"""
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:55]:
```suggestion
    """Test wrapper for Pandas ``.str.get()`` method"""
```

File: tests/functions/test_process_text.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            ],
+            "numbers": [1, 2, 3, 4],
+        }
+    )
+
+    assert_frame_equal(result, expected)
+
+
+def test_str_cat():
+    "Test Pandas string cat method"
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:36]:
```suggestion
    "Test wrapper for Pandas ``.str.cat()`` method"
```

File: tests/functions/test_process_text.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
+import numpy as np
+import pandas as pd
+import pytest
+from pandas.testing import assert_frame_equal
+
+from janitor.functions import process_text
+
+
+def test_str_split():
+    "Test Pandas string split method"
```
Comment (ericmjl) [inline tests/functions/test_process_text.py:10]:
```suggestion
    "Test wrapper for Pandas ``.str.split()`` method"
```

File: tests/functions/test_process_text.py (hunk 8 above)
Comment (hectormz) [reply tests/functions/test_process_text.py:112]: In this situation, I like to make an over-the-top obviously invalid function like:
* `invalid_function`
* `nonexistent_function`

Just to make it very obvious for others later

**Comment by samukweku** [comment]
@ericmjl @HectorM14 Thanks! got some guidance from @VPerrollaz script, so it's a community effort :) I will add more tests, and possibly check Stack Overflow or some blogs for some edge cases. Not familiar with JanitorError, so cant say much on that; I default to Pandas exception handling. As always, more feedback are welcome.

**Comment by hectormz** [comment]
@samukweku You don't *have* to add new tests. I was just curious if they were picked for a reason, or if there is an ideal set that covers different issues.

Do all the string methods return a series?

**Comment by samukweku** [comment]
@HectorM14 , yes, all the string methods return a series, same as in Pandas. The tests were picked at random, just testing for those that needed extra parameters and just one test for string methods that required no parameters

**Comment by VPerrollaz** [comment]
Looks excellent. The api feels right and it is well documented.

File: janitor/functions.py [hunk 14]
```diff
@@ -4001,3 +4002,89 @@ def expand_grid(
...
+        # text       |   code
+        # ragnar     |    1
+        # sammywemmy |    2
+        # ginger     |    3
+
+        #For string methods with parameters, simply pass the arguments :
+        df.process_text(column = "text",
+                        string_function = "extract",
+                        pat = r"(ag)",
+                        flags = re.IGNORECASE)
```
Comment (VPerrollaz) [inline janitor/functions.py:4042]:
```python
df.process_text(
    column = "text",
    string_function = "extract",
    pat = r"(ag)",
    flags = re.IGNORECASE
)
```

**Comment by ericmjl** [comment]
@samukweku I restarted the build. Let's see whether it fails again.

**Comment by samukweku** [comment]
@ericmjl , still the same issue with CI build

**Comment by ericmjl** [comment]
> @ericmjl , still the same issue with CI build

Really weird. I don't really know why that's happening, but I figure it might be due to the conda solver taking an insane amount of time to finish up. I recently tried the mamba solver, and I think it might be a better way to go.

I'll put up an issue for adding the mamba solver to the CI system.

Meanwhile, can you check to see whether you are able to restart individual jobs? I figure that you should be able to, since you have commit rights to the repository.

**Comment by ericmjl** [comment]
Actually, @samukweku, I think you can sneak in your first infrastructure changes here. There are two things we probably need to change.

Firstly, can you modify the .travis.yml to add in `conda install -c conda-forge mamba` right after https://github.com/ericmjl/pyjanitor/blob/dev/.travis.yml#L19? This guarantees that we install the mamba environment solver.

Secondly, can you change the following conda install commands in the .travis.yml file?
- `conda env create -f environment-dev.yml` --> `mamba env create -f environment-dev.yml`
- `conda install python=$PYTHON_VERSION` --> `mamba install python=$PYTHON_VERSION -n pyjanitor-dev`

Thirdly, can you comment out the devcontainer build line? It is `docker build -t ericmjl/pyjanitor:devcontainer -f .devcontainer/Dockerfile .` We don't need it to run, as I will switch it over to building on Dockerhub directly.

**Comment by hectormz** [comment]
mamba IRL! 🤩

File: janitor/functions.py [hunk 15]
```diff
@@ -4001,3 +4004,104 @@ def expand_grid(
...
+def process_text(
+    df: pd.DataFrame,
+    column: str,
+    string_function: str,
+    *args: str,
+    **kwargs: str,
+) -> pd.DataFrame:
+    """
+    Applies a Pandas string method to an existing column,
+    and returns a dataframe.
```
Comment (hectormz) [inline janitor/functions.py:4019]:
```suggestion
    Apply a Pandas string method to an existing column and return a dataframe.
    
```

File: janitor/functions.py [hunk 16]
```diff
@@ -4001,3 +4004,104 @@ def expand_grid(
...
+    This function aims to make string cleaning easy, while chaining,
+    by simply passing the string method name to the
+    ``process_text`` function.
+    Note that this modifies an existing column, and should not be
+    used to create a new column.
```
Comment (hectormz) [inline janitor/functions.py:4024]:
```suggestion
    This function aims to make string cleaning easy, while chaining,
    by simply passing the string method name to the ``process_text`` function.
    Note that this modifies an existing column,
    and should not be used to create a new column.
```

File: tests/functions/test_process_text.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+    expected = df.copy()
+    expected["text"] = expected["text"].str.get(1)
+    result = df.process_text(column="text", string_function="get", i=-1)
+
+    assert_frame_equal(result, expected)
+
+
+def test_str_lower():
+    """Test string conversion to lowercase using ``.str.lower()``"""
```
Comment (hectormz) [inline tests/functions/test_process_text.py:52]:
```suggestion
    """Test string conversion to lowercase using ``.str.lower()``."""
```

File: tests/functions/test_process_text.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    result = df.process_text(column="names", string_function="lower")
+
+    assert_frame_equal(result, expected)
+
+
+def test_str_wrong():
+    """
+    Test that string_function that is not a Pandas string method
+    actually errors out.
+    """
```
Comment (hectormz) [inline tests/functions/test_process_text.py:80]:
```suggestion
    """Test that an invalid Pandas string method raises an exception."""
```

File: tests/functions/test_process_text.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    )
+    with pytest.raises(KeyError):
+        df.process_text(column="text", string_function="ragnar")
+
+
+def test_str_wrong_parameters():
+    """
+    Test that wrong parameters for a Pandas string method
+    actually errors out.
+    """
```
Comment (hectormz) [inline tests/functions/test_process_text.py:92]:
```suggestion
    """Test that invalid argument for Pandas string method raises an error."""
```

File: tests/functions/test_process_text.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+def test_str_wrong():
+    """
+    Test that string_function that is not a Pandas string method
+    actually errors out.
+    """
+    df = pd.DataFrame(
+        {"text": ["ragnar", "sammywemmy", "ginger"], "code": [1, 2, 3]}
+    )
+    with pytest.raises(KeyError):
+        df.process_text(column="text", string_function="ragnar")
```
Comment (hectormz) [inline tests/functions/test_process_text.py:85]:
```suggestion
        df.process_text(column="text", string_function="invalid_function")
```

**Comment by hectormz** [comment]
If the conda/mamba solving is fixed, I just added some requested changes to tighten up the docstrings and it's all good to go @samukweku !

**Comment by samukweku** [comment]
Yes, problem fixed. @hectormz @ericmjl how did you know what to fix for the build error? New to this, so questions might be silly. When I checked the view log, the only error I noticed was ``deploy.provider : unknown value pages:git``.

**Comment by samukweku** [comment]
@ericmjl Also, how do I fix the deepsource error?

**Comment by ericmjl** [comment]
@samukweku when I looked at the build log, I saw that only that particular build stalled at the docker container build step in which the conda env solving piece looked slow.

As such, I had the following hypotheses:

- Something went wrong with conda
- Something went wrong with docker build

At first, I thought it was conda stalling. In writing the response to you, I then realized, oh, the first conda step actually worked, so maybe the docker build was the real issue. But I wasn’t sure how to debug what’s going on in a Docker container build on a remote machine, so I thought, let’s just disable it for now, since it’s not 100% crucial for the package. It’s merely helpful for development.

That was my thought process! Hopefully that’s informative for you too :).

**Comment by ericmjl** [comment]
As for the current errors, let’s see.

For DeepSource: does the deep source config file (`.deepsource.toml`) exist in your fork?

For the Azure builds, looks like one was an HTTP error. That just needs a restart. I think I have added you to the pipelines, @samukweku. Can you see if you can restart the build? (You might need to login to Azure again.)

**Comment by samukweku** [comment]
Didn't have to do anything except push new commit. Fingers crossed.

**Comment by ericmjl** [comment]
Everything looks good to me. @hectormz would you like to do the honors?

**Comment by hectormz** [comment]
@ericmjl  will do! just fixed one last thing in .rst's. Will wait for the CI to finish and then will merge!

**Comment by ericmjl** [comment]
:tada:
